source/common/network/utility.cc

```diff
@@ -541,19 +549,56 @@ Api::IoCallUint64Result Utility::readFromSocket(Network::Socket& socket,
 
   RELEASE_ASSERT(output.peer_address_ != nullptr,
                  fmt::format("Unable to get remote address for fd: {}, local address: {} ",
-                             socket.ioHandle().fd(), socket.localAddress()->asString()));
+                             handle.fd(), local_address.asString()));
 
   // Unix domain sockets are not supported
   RELEASE_ASSERT(output.peer_address_->type() == Address::Type::Ip,
                  fmt::format("Unsupported remote address: {} local address: {}, receive size: "
                              "{}",
-                             output.peer_address_->asString(), socket.localAddress()->asString(),
+                             output.peer_address_->asString(), local_address.asString(),
                              result.rc_));
   udp_packet_processor.processPacket(std::move(output.local_address_),
                                      std::move(output.peer_address_), std::move(buffer),
                                      receive_time);
   return result;
 }
 
+Api::IoErrorPtr Utility::readPacketsFromSocket(IoHandle& handle,
+                                               const Address::Instance& local_address,
+                                               UdpPacketProcessor& udp_packet_processor,
+                                               TimeSource& time_source, uint32_t& packets_dropped) {
+  do {
+    const uint32_t old_packets_dropped = packets_dropped;
+    const MonotonicTime receive_time = time_source.monotonicTime();
+    Api::IoCallUint64Result result = Utility::readFromSocket(
+        handle, local_address, udp_packet_processor, receive_time, &packets_dropped);
+
+    if (!result.ok()) {
+      // No more to read or encountered a system error.
+      return std::move(result.err_);
+    }
+
+    if (result.rc_ == 0) {
+      // TODO(conqerAtapple): Is zero length packet interesting? If so add stats
+      // for it. Otherwise remove the warning log below.
+      ENVOY_LOG_MISC(trace, "received 0-length packet");
+    }
+
+    if (packets_dropped != old_packets_dropped) {
+      // The kernel tracks SO_RXQ_OVFL as a uint32 which can overflow to a smaller
+      // value. So as long as this count differs from previously recorded value,
+      // more packets are dropped by kernel.
+      const uint32_t delta =
+          (packets_dropped > old_packets_dropped)
+              ? (packets_dropped - old_packets_dropped)
+              : (packets_dropped + (std::numeric_limits<uint32_t>::max() - old_packets_dropped) +
+                 1);
+      // TODO(danzh) add stats for this.
+      ENVOY_LOG_MISC(
+          debug, "Kernel dropped {} more packets. Consider increase receive buffer size.", delta);
+    }
+  } while (true);
```

Comment:
@mattklein123 this loop runs until we run out of packets; do you think we should have some bound on the loop to ensure fairness across epoll events? This may be a possible DoS vector once UDP becomes a thing.

Reply:
Yes @danzh2010 and I have already discussed thi and there are various TODOs in the code. Someone will be doing this as a follow up once the main support for UDP and QUIC merges.